### ISSUE TITLE
Exception in  IClassListener.onBeforeClass() causes ignore tests instead of fail.

### ISSUE THREAD

**Issue by mechanik0s** [testng-team/testng#2295]
### TestNG Version
7.1.1
### Expected behavior
Exception thrown in `onBeforeClass()` method should be handled like exception thrown in `@BeforeClass` method and cause test fail.

### Actual behavior
Exception thrown in `onBeforeClass()` method causes ignore test

### Is the issue reproductible on runner?

- [ ] Shell
- [ ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x] IntelliJ
- [ ] NetBeans

### Test case sample
```java
@Listeners({
        ClassListener.class
})
public class A {
    @BeforeClass
    public void beforeClass() {
        System.out.println("A::@BeforeClass");
        //throw new RuntimeException();   Uncomment this line and see that causes mark A class as failed and it's good
    }

    @Test
    public void testA() {
        System.out.println("A::@Test");
    }
}
```
```java
public class ClassListener implements IClassListener {
    @Override
    public void onBeforeClass(ITestClass testClass) {
        System.out.println("ClassListener::OnBeforeClass");
        throw new RuntimeException();   //Cause mark A class as ignored and it's bad
    }

    @Override
    public void onAfterClass(ITestClass testClass) {
        System.out.println("ClassListener::OnAfterClass");
    }
}
```

### Possible solutions 
1. Handle exceptions in `org.testng.internal.TestMethodWorker#invokeBeforeClassMethods`
2. Pass test context to IClassListener and give opportunity to enforce test's fail in IClassListener

**Comment by krmahadevan** [comment]
@mechanik0s do you want to raise a PR to address this issue since it looks like you have analysed the problem as well.

**Comment by krmahadevan** [comment]
@juherr - How do you suggest that we handle this?
Currently TestNG doesn't handle any exceptions from any listeners. So the onus is on the listeners to ensure that they dont throw any exceptions. In this case, the listener throws an exception and so further execution gets halted since we dont move forward. I am in for closing this issue with resolution as **working as designed**

**Comment by juherr** [comment]
I agree with the issue. I don't see any reason that configuration methods and listeners have different behaviors, do you?

**Comment by krmahadevan** [comment]
@juherr - Here's my point of view on this.

 Listeners are never part of a test per se. They are basically callbacks that TestNG provides to let a user deal with events as and when they occur.
Configuration methods on the other hand are part of the test, because they perform some sort of a `setup` that is needed by the Test.

So failures in configurations should affect the tests and TestNG should handle them to translate them into the next set of actions to be done on the test (Skip the test for e.g.,)

Listeners on the other hand are callbacks. They shouldn't be throwing exceptions in the first place. So why should TestNG handle them and have their exceptions affect the outcome of the tests?

We can either gobble all exceptions and just warn the user (which kind of makes it impossible for the user to put a recovery mechanism on their side, or we can explicitly document that listeners if they throw exceptions will abort the tests.

Its similar to our reporting listener (`IReporter` implementation) throwing an exception and that causing the tests to fail (right now it doesn't happen coz the listener gets invoked later). But I hope that analog explains my point.

Summarising: TestNG is working as designed here. Listeners are not expected to throw exceptions. They either deal with the exceptions or TestNG aborts execution.

@cbeust - Your thoughts ?

**Comment by krmahadevan** [comment]
@juherr  - What do you suggest we do with this ?

**Comment by juherr** [comment]
@krmahadevan 
I keep my previous point of view: `IClassListener.onBeforeClass()` and `@BeforeClass` are 2 different ways to define the same things and should have the exact same behavior.
From an implementation point of view, I'd love to have the same workflow for both (where `@BeforeClass` could be wired and translated into an internal listener).

I remember a previous discussion where we said that the configuration methods are not part of the test and exceptions into them should fail the run because the users have to manage the exceptions and the test status themself.

In other words, I think exceptions in a listener or a configuration method should not be allowed. If any, the run should fail and not set any test status (the documentation should highlight this).

Thought?

**Comment by juherr** [comment]
AI triage verdict: **probably still relevant**.

`TestMethodWorker` still invokes `IClassListener.onBeforeClass()` directly, without translating listener exceptions into configuration/test failures. The surrounding execution path has changed since this report: `GraphOrchestrator`/`TestTaskExecutor` now retain and log worker throwables (see #3243), so the exact observable outcome from 7.1.1 cannot be assumed unchanged. I found no current regression test for an `IClassListener.onBeforeClass()` that throws, and no open or merged PR addressing #2295 directly.

The original reproducer was **not executed** in this triage pass.

Minimal regression test: run a class with one `@Test` and an `IClassListener` whose `onBeforeClass()` throws, then assert the resulting TestNG run status and the test/configuration result maps. This should first establish today's behavior. Only after that should production code be changed to implement the intended failure semantics.
